ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/state/cache/PendingBlockCache.java

```diff
@@ -34,25 +37,39 @@ public PendingBlockCache(final long cacheSizePerPeer) {
    *
    * @return the previous value associated with the specified key, or {@code null} if there was no
    *     mapping for the hash
-   * @throws IndexOutOfBoundsException if the limit of the number of blocks has been reached for
-   *     this nodeId
    */
   @Override
   public ImmutablePendingBlock putIfAbsent(
-      final Hash hash, final ImmutablePendingBlock pendingBlock) throws IndexOutOfBoundsException {
-    if (getPeerWeight(pendingBlock.nodeId()) >= cacheSizePerPeer) {
-      throw new IndexOutOfBoundsException();
+      final Hash hash, final ImmutablePendingBlock pendingBlock) {
+    final ImmutablePendingBlock foundBlock = super.putIfAbsent(hash, pendingBlock);
+    if (foundBlock == null) {
+      removeLowestPriorityBlockWhenCacheFull(pendingBlock.nodeId());
     }
-    return super.putIfAbsent(hash, pendingBlock);
+    return foundBlock;
   }
 
   /**
-   * Returns the number of pending blocks from a node that are stored in the cache
+   * Removes the lowest priority block if a peer has reached the cache limit it is allowed to use
+   * The highest priority blocks are those that are lowest in block height and then higher priority
+   * if they were sent more recently.
    *
-   * @param nodeId the peer ID
-   * @return the number of elements in the cache coming from this node
+   * @param nodeId id of the peer
    */
-  private long getPeerWeight(final Bytes nodeId) {
-    return values().stream().filter(value -> value.nodeId() == nodeId).count();
+  private void removeLowestPriorityBlockWhenCacheFull(final Bytes nodeId) {
+    final List<ImmutablePendingBlock> blockByNodeId =
+        values().stream().filter(value -> value.nodeId() == nodeId).collect(Collectors.toList());
+    if (blockByNodeId.size() > cacheSizePerPeer) {
+      blockByNodeId.stream()
+          .min(getComparatorByBlockNumber().reversed().thenComparing(getComparatorByTimeStamp()))
```

Comment:
I think I'm missing something, why do we need the reversed? Don't we want to keep the lowest block numbers?

Reply:
Yes The minimum will be done on the result after the two comparators.

1) Sorting by block number is by ascending 
2) Reversing order (now it's descending)
3) Sorting by timestamp is also by ascending .
4) Get the minimum 

Otherwise the result would be the smallest number block with the smallest timestamp

Live sample code : [sample code to execute](https://www.jdoodle.com/iembed/v0/2rv)

```java
final List<Block> list = new ArrayList<>();
list.add(new Block(4,111111));
list.add(new Block(1,2222222));
list.add(new Block(4,2222222));
list.add(new Block(3,3333333));
```

**Without reversed :**
Block{number=1, timestamp=2222222}

**With reversed:**
Block{number=4, timestamp=111111}

Reply:
Right, that's what I thought we wanted. If there's a reorg and a node is sending us new blocks we want to make sure that we keep the ones that are lower because they're the ones we're going to need to import sooner.

Reply:
They're also the ones we can throw out sooner when they fall out of our import range.

Reply:
yes that's what this code does.  we get the block that we want to replace

Reply:
Step by step : 

At the beginning :
```java
list.add(new Block(4,111111));
list.add(new Block(1,2222222));
list.add(new Block(4,2222222));
list.add(new Block(3,3333333));
```

1- After getComparatorByBlockNumber:
```java
Block{number=1, timestamp=2222222}
Block{number=3, timestamp=3333333}
Block{number=4, timestamp=111111}
Block{number=4, timestamp=2222222}
```

2- After reversed :
```java
Block{number=4, timestamp=2222222}
Block{number=4, timestamp=111111}
Block{number=3, timestamp=3333333}
Block{number=1, timestamp=2222222}
```

3- After getComparatorByTimeStamp :
```java
Block{number=4, timestamp=111111}
Block{number=4, timestamp=2222222}
Block{number=3, timestamp=3333333}
Block{number=1, timestamp=2222222}
```
4- After min
```java
Block{number=4, timestamp=111111}
```

-> Replace `Block{number=4, timestamp=111111}` (lowest priority)

Reply:
Oh yeah, you're right :)